calyx/src/passes/compile_ref.rs

```diff
@@ -13,8 +13,34 @@ use std::rc::Rc;
 /// 2. Inline all the ports of the ref cells to the component signature
 /// 3. Remove all the ref cell mappings from the invoke statement
 /// 4. Inline all the mappings of ports to the invoke signature
+
+/// Map for storing added ports for each ref cell
+/// level of Hashmap represents:
+/// HashMap<-component name-, Hashmap<-cell name-, HashMap<-port name-, port>>>;
+type RefPortMap =
+    HashMap<ir::Id, HashMap<ir::Id, HashMap<ir::Id, RRC<ir::Port>>>>;
```

Comment:
Try using `port.canonical` to get `(cell_name, port_name)` tuple which can be used to index into the hashmap